backend/protocol_rpc/endpoints.py

```diff
@@ -401,14 +407,24 @@ def call(
         ),
         leader_receipt=None,
         msg_handler=msg_handler.with_client_session(get_client_session_id()),
-        contract_snapshot_factory=partial(ContractSnapshot, session=session),
     )
 
-    return node.get_contract_data(
-        code=contract_account["data"]["code"],
-        state=contract_account["data"]["state"],
+    receipt = await node.get_contract_data(
+        from_address="0x" + "00" * 20,
         calldata=decoded_data.calldata,
     )
+    if receipt.error is not None or receipt.returned is None:
+        return receipt.to_dict()
+    try:
+        import backend.node.genvm.origin.host_fns as genvm_consts
+
+        if receipt.returned[0] == genvm_consts.ResultCode.ROLLBACK:
+            return "Rollback: " + str(receipt.returned[1:], "utf-8")
+        elif receipt.returned[0] == genvm_consts.ResultCode.RETURN:
+            return genvm_calldata.to_str(genvm_calldata.decode(receipt.returned[1:]))
+    except:
+        pass
+    return receipt.to_dict()
```

Comment:
are these 2 paths compatible from a client's perspective?

Reply:
This is an awful workaround for the fact that simulator doesn't properly support returns:
- write methods can return as well
- read methods should go through a consensus

This line is sort of incorrect `genvm_calldata.to_str(genvm_calldata.decode(receipt.returned[1:]))`, because we should not be decoding anything at all on the backend and we should just return the receipt to the end user (frontend) and let it handle it

However, I think it is out of scope of this PR

Reply:
and answering the original question: no, they are not

Reply:
@kp2pml30, could you please explain why the read method should go through consensus?

Reply:
I don't think that I am qualified to answer it or create an issue. As far as I understand this `call` tries to emulate `eth_call`

initially geth [didn't support errors](https://github.com/ethereum/go-ethereum/issues/19027), but it appears they [fixed it](https://github.com/ethereum/go-ethereum/pull/21083)

I think that valid path would be to return `returned` on success (without decoding and dumping to string), because it is what eth does. And on error return "something" (right now it is dictionary form of `Receipt`)

Reply:
see #621